### PR TITLE
Add `no_std` support to `bevy_state`

### PR DESCRIPTION
**Pull request by bushrat011899**
# Objective

- Contributes to #15460

## Solution

- Added the following features:
  - `std` (default)
  - `portable-atomic`
  - `critical-section`

## Testing

- CI

## Notes

- `portable-atomic`, and `critical-section` are shortcuts to enable the relevant features in dependencies, making the usage of this crate on atomically challenged platforms possible and simpler.
- This PR is blocked until #17027 is merged (as it depends on fixes for the `once!` macro). Once merged, the change-count for this PR should reduce.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]

**Review by BenjaminBrienen — APPROVED** [review]
Easy one!

**Comment by alice-i-cecile** [comment]
@bushrat011899 ping me to merge this once merge conflicts are resolved please :)

**Comment by bushrat011899** [comment]
@alice-i-cecile Conflicts resolved! It's a shame the git algorithm can't handle the way I add tests to that `compile-check-no-std` command, but at least it's an easy resolve.

**Review by BenjaminBrienen — APPROVED** [review]

**Pull request merged**
